File: adapter/js/zoom-parser.js
'use strict';

/**
 * Zoom CSV Parser
 *
 * Parses Zoom "meeting participant details" reports into structured data.
 *
 * Handles:
 *   - BOM marker
 *   - Quoted CSV fields (dates contain commas)
 *   - Duplicate header "Durata (minuti)" (meeting-level vs participant-level)
 *   - Multiple meetings per file (separated by blank lines)
 *   - Host filtering (Guest = "No")
 */
const ZoomParser = {

    /**
     * Parse a Zoom CSV file into an array of meeting objects.
     *
     * @param {string} csvText - Raw file content
     * @returns {{ meetings: Array, warnings: Array<string> }}
     */
    parse(csvText) {
        const warnings = [];

        const clean = csvText.replace(/^\uFEFF/, '');
        const lines = clean.split('\n');

        if (lines.length < 2) {
            throw new Error('Il file CSV deve contenere almeno intestazioni e una riga');
        }

        const rawHeaders = this._parseLine(lines[0]);
        const headers = this._deduplicateHeaders(rawHeaders);

        const meetings = [];
        let currentRows = [];

        for (let i = 1; i < lines.length; i++) {
            const trimmed = lines[i].trim();

            if (!trimmed) {
                if (currentRows.length > 0) {
                    const m = this._buildMeeting(currentRows, warnings);
                    if (m) meetings.push(m);
                    currentRows = [];
                }
                continue;
            }

            const values = this._parseLine(trimmed);
            if (values.length >= headers.length) {
                const row = {};
                headers.forEach((h, idx) => {
                    row[h] = (values[idx] || '').trim();
                });
                currentRows.push(row);
            } else {
                warnings.push(`Riga ${i + 1}: colonne non corrispondenti (${values.length}/${headers.length}), saltata`);
            }
        }

        // Last meeting (file may not end with blank line)
        if (currentRows.length > 0) {
            const m = this._buildMeeting(currentRows, warnings);
            if (m) meetings.push(m);
        }

        return { meetings, warnings };
    },

    // ------------------------------------------------------------------
    // Private
    // ------------------------------------------------------------------

    _buildMeeting(rows, warnings) {
        if (rows.length === 0) return null;

        const first = rows[0];
        const startTime = this._parseDateTime(first['Ora di inizio']);
        const endTime = this._parseDateTime(first['Ora di fine']);

        if (!startTime || !endTime) {
            warnings.push(`Meeting "${first['Argomento'] || '?'}": date non valide, saltato`);
            return null;
        }

        const meeting = {
            course: first['Argomento'] || '',
            meetingId: first['ID'] || '',
            organizer: first['Nome organizzatore'] || '',
            organizerEmail: first['E-mail organizzatore'] || '',
            startTime,
            endTime,
            durationMinutes: (endTime - startTime) / (1000 * 60),
            segments: []
        };

        for (const row of rows) {
            const rawName = row['Nome (nome originale)'] || '';
            const email = row['E-mail'] || '';
            const guestField = (row['Guest'] || '').toLowerCase();

            // Skip non-guests (hosts / organizers)
            if (guestField !== 'sì' && guestField !== 'si') {
                continue;
            }

            const joinTime = this._parseDateTime(row['Ora di ingresso']);
            const leaveTime = this._parseDateTime(row['Ora di uscita']);

            if (!joinTime || !leaveTime) {
                warnings.push(`Meeting "${meeting.course}": segmento con date non valide per "${rawName}", saltato`);
                continue;
            }

            const { firstName, lastName } = this._splitName(rawName);

            meeting.segments.push({
                fullName: rawName,
                firstName,
                lastName,
                email,
                joinTime,
                leaveTime,
                durationMinutes: (leaveTime - joinTime) / (1000 * 60)
            });
        }

        return meeting;
    },

    /**
     * Parse a CSV line handling quoted fields.
     */
    _parseLine(line) {
        const result = [];
        let current = '';
        let inQuotes = false;

        for (let i = 0; i < line.length; i++) {
            const ch = line[i];
            if (ch === '"') {
                if (inQuotes && line[i + 1] === '"') {
                    current += '"';
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch === ',' && !inQuotes) {
                result.push(current);
                current = '';
            } else {
                current += ch;
            }
        }
        result.push(current);
        return result;
    },

    /**
     * Make duplicate header names unique: "X", "X" → "X", "X_2"
     */
    _deduplicateHeaders(headers) {
        const counts = {};
        return headers.map(h => {
            const t = h.trim();
            if (!counts[t]) { counts[t] = 1; return t; }
            counts[t]++;
            return `${t}_${counts[t]}`;
        });
    },

    /**
     * Parse Zoom date string "MM/DD/YYYY HH:MM:SS AM/PM" → Date
     */
    _parseDateTime(str) {
        if (!str) return null;
        const m = str.match(/(\d{2})\/(\d{2})\/(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})\s+(AM|PM)/i);
        if (!m) return null;

        let [, month, day, year, hours, minutes, seconds, period] = m;
        hours = parseInt(hours);
        if (period.toUpperCase() === 'PM' && hours !== 12) hours += 12;
        if (period.toUpperCase() === 'AM' && hours === 12) hours = 0;

        return new Date(
            parseInt(year), parseInt(month) - 1, parseInt(day),
            hours, parseInt(minutes), parseInt(seconds)
        );
    },

    /**
     * Split "Daniel Lo Russo" → { firstName: "Daniel", lastName: "Lo Russo" }
     * Uses first word as first name, rest as last name.
     */
    _splitName(fullName) {
        const clean = fullName
            .replace(/\(Host\)/gi, '')
            .replace(/\([^)]*\)/g, '')
            .trim();

        const parts = clean.split(/\s+/);
        if (parts.length === 0) return { firstName: '', lastName: '' };
        if (parts.length === 1) return { firstName: this._cap(parts[0]), lastName: '' };

        return {
            firstName: this._cap(parts[0]),
            lastName: parts.slice(1).map(p => this._cap(p)).join(' ')
        };
    },

    /** Capitalize first letter: "luca" → "Luca" */
    _cap(word) {
        if (!word) return '';
        return word.charAt(0).toUpperCase() + word.slice(1);
    }
};

File: backend/attendance_normalization/zoom_parser.py
"""Zoom CSV parser migrated from ``attendance/adapter/js/zoom-parser.js``."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
import re

from .aggregator import ZoomMeeting, ZoomSegment


@dataclass(frozen=True)
class ZoomParseResult:
    meetings: list[ZoomMeeting]
    warnings: list[str]


ZOOM_DATETIME_PATTERN = re.compile(
    r"(\d{2})\/(\d{2})\/(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})\s+(AM|PM)",
    re.IGNORECASE,
)


def parse_zoom_csv_text(csv_text: str) -> ZoomParseResult:
    warnings: list[str] = []
    clean = csv_text.removeprefix("\ufeff")
    lines = clean.splitlines()

    if len(lines) < 2:
        raise ValueError("Il file CSV deve contenere almeno intestazioni e una riga")

    raw_headers = _parse_csv_line(lines[0])
    headers = _deduplicate_headers(raw_headers)

    meetings: list[ZoomMeeting] = []
    current_rows: list[dict[str, str]] = []

    for index, line in enumerate(lines[1:], start=2):
        trimmed = line.strip()

        if not trimmed or not trimmed.replace(",", ""):
            if current_rows:
                meeting = _build_meeting(current_rows, warnings)
                if meeting is not None:
                    meetings.append(meeting)
                current_rows = []
            continue

        values = _parse_csv_line(trimmed)
        if len(values) >= len(headers):
            row = {
                header: (values[position] or "").strip()
                for position, header in enumerate(headers)
            }
            current_rows.append(row)
        else:
            warnings.append(
                f"Riga {index}: colonne non corrispondenti ({len(values)}/{len(headers)}), saltata"
            )

    if current_rows:
        meeting = _build_meeting(current_rows, warnings)
        if meeting is not None:
            meetings.append(meeting)

    return ZoomParseResult(meetings=meetings, warnings=warnings)


def parse_zoom_csv_file(path: str) -> ZoomParseResult:
    with open(path, "r", encoding="utf-8-sig", newline="") as handle:
        return parse_zoom_csv_text(handle.read())


def _build_meeting(rows: list[dict[str, str]], warnings: list[str]) -> ZoomMeeting | None:
    if not rows:
        return None

    first = rows[0]
    start_time = _parse_zoom_datetime(first.get("Ora di inizio", ""))
    end_time = _parse_zoom_datetime(first.get("Ora di fine", ""))

    if start_time is None or end_time is None:
        warnings.append(
            f'Meeting "{first.get("Argomento", "?")}": date non valide, saltato'
        )
        return None

    segments: list[ZoomSegment] = []
    for row in rows:
        raw_name = row.get("Nome (nome originale)", "")
        email = row.get("E-mail", "")
        guest_field = row.get("Guest", "").lower()

        if guest_field not in {"sì", "si"}:
            continue

        join_time = _parse_zoom_datetime(row.get("Ora di ingresso", ""))
        leave_time = _parse_zoom_datetime(row.get("Ora di uscita", ""))

        if join_time is None or leave_time is None:
            warnings.append(
                f'Meeting "{first.get("Argomento", "")}": segmento con date non valide per "{raw_name}", saltato'
            )
            continue

        first_name, last_name = _split_name(raw_name)
        segments.append(
            ZoomSegment(
                full_name=raw_name,
                first_name=first_name,
                last_name=last_name,
                email=email,
                join_time=join_time,
                leave_time=leave_time,
            )
        )

    return ZoomMeeting(
        course=first.get("Argomento", ""),
        meeting_id=first.get("ID", ""),
        start_time=start_time,
        end_time=end_time,
        duration_minutes=(end_time - start_time).total_seconds() / 60,
        segments=segments,
    )


def _parse_csv_line(line: str) -> list[str]:
    result: list[str] = []
    current = []
    in_quotes = False
    index = 0

    while index < len(line):
        char = line[index]
        if char == '"':
            if in_quotes and index + 1 < len(line) and line[index + 1] == '"':
                current.append('"')
                index += 1
            else:
                in_quotes = not in_quotes
        elif char == "," and not in_quotes:
            result.append("".join(current))
            current = []
        else:
            current.append(char)
        index += 1

    result.append("".join(current))
    return result


def _deduplicate_headers(headers: list[str]) -> list[str]:
    counts: dict[str, int] = {}
    deduplicated: list[str] = []
    for header in headers:
        trimmed = header.strip()
        counts[trimmed] = counts.get(trimmed, 0) + 1
        if counts[trimmed] == 1:
            deduplicated.append(trimmed)
        else:
            deduplicated.append(f"{trimmed}_{counts[trimmed]}")
    return deduplicated


def _parse_zoom_datetime(value: str) -> datetime | None:
    if not value:
        return None

    match = ZOOM_DATETIME_PATTERN.match(value)
    if match is None:
        return None

    month, day, year, hours, minutes, seconds, period = match.groups()
    parsed_hours = int(hours)
    if period.upper() == "PM" and parsed_hours != 12:
        parsed_hours += 12
    if period.upper() == "AM" and parsed_hours == 12:
        parsed_hours = 0

    return datetime(
        int(year),
        int(month),
        int(day),
        parsed_hours,
        int(minutes),
        int(seconds),
    )


def _split_name(full_name: str) -> tuple[str, str]:
    clean = re.sub(r"\(Host\)", "", full_name, flags=re.IGNORECASE)
    clean = re.sub(r"\([^)]*\)", "", clean).strip()
    parts = clean.split()
    if not parts:
        return "", ""
    if len(parts) == 1:
        return _cap(parts[0]), ""
    return _cap(parts[0]), " ".join(_cap(part) for part in parts[1:])


def _cap(word: str) -> str:
    if not word:
        return ""
    return word[0].upper() + word[1:]
